builtin/common/vector.lua

```diff
@@ -1,58 +1,96 @@
+--[[
+Vector helpers
+
+Note: The functions must be able to accept old vectors that had no metatables and
+      looked like this: {x = 42, y = 13, z = 0}
+]]
 
 vector = {}
 
+local metatable = {}
+
+function metatable.__index(v, key)
+	if type(key) == "number" then
+		return v[key]
+	elseif key == "x" then
+		return v[1]
+	elseif key == "y" then
+		return v[2]
+	elseif key == "z" then
+		return v[3]
+	end
+	local func = vector[key]
+	if func then
+		return func
+	else
+		error("Invalid key for vector", 2)
+	end
+end
+
+function metatable.__newindex(v, key, value)
+	if type(key) == "number" then
+		v[key] = value
+	elseif key == "x" then
+		v[1] = value
+	elseif key == "y" then
+		v[2] = value
+	elseif key == "z" then
+		v[3] = value
+	else
+		error("Invalid key for vector", 2)
```

Comment:
Not being able to set foreign keys in a vector is a possible backwards compatibility break.

Reply:
Done.